### PR TITLE
Add Pretty instance for log lines.

### PR DESCRIPTION
**Pull request by zmrocze**
- new `LogLine` newtype with pretty instance
 - new `sufficientLogLevel`
 - fixes bug where logContext and logLevel were printed twice

Point of this pr is only to export existing logs pretty printing and checking if log level is sufficient, to be then reused in plutip (we use `Show` now unnecessarily). This needed a `LogLine` newtype for Pretty instance.

I didn't define `sufficientLogLevel` to be `Ord` instance for `LogLevel` because it wouldn't be transitive, but can change it.

As a side-effect it fixes my previous mistake where log level and context would be printed twice.

### CODE REVIEW

File: src/BotPlutusInterface/Types.hs [hunk 1]
```diff
@@ -277,6 +296,23 @@ instance Pretty LogLevel where
...
+ then message should be displayed with this log level setting.
+-}
+sufficientLogLevel :: LogLevel -> LogLevel -> Bool
+sufficientLogLevel logLevelSetting msgLogLvl =
+  msgLogLvl `constrLEq` logLevelSetting -- the log is important enough
+    && not (null intersectLogTypes) -- log is of type we're interested in
+  where
+    intersectLogTypes = ltLogTypes logLevelSetting `intersect` (ltLogTypes msgLogLvl <> [AnyLog])
+
+    toOrd (Error _) = 0 :: Int
```
Comment (IAmPara0x) [inline src/BotPlutusInterface/Types.hs:309]: A bit of bike-shedding, I think we can eliminate the need to use toOrd by using `Data.Data` and `Enum`. Here's the complete code for that:
```
deriving stock instance Data LogType
deriving stock instance Data LogLevel

instance Enum LogLevel where
  fromEnum = (\a -> a - 1) . constrIndex . toConstr
  toEnum   = fromConstrB field . indexConstr (dataTypeOf $ Notice []) . (+ 1)

    where

      field :: forall a. Data a => a
      field = case eqT :: Maybe (a :~: [LogType]) of
          Just Refl -> [AnyLog]
          Nothing   -> error "Expected a value of type LogType."
```
in this way, if we add new data constructor for `LogLevel` we will not have to modify `toOrd`, and it's `Enum` instance will also get automatically handled for us.

The new `toOrd` will be:
```
toOrd a b = fromEnum a <= fromEnum b
```
For this you will need the following extension `{-# LANGUAGE DeriveDataTypeable #-}`.

It's just an idea, if you like it feel free to add it.

**Review by IAmPara0x — APPROVED** [review]

File: src/BotPlutusInterface/Types.hs (hunk 1 above)
Comment (zmrocze) [reply src/BotPlutusInterface/Types.hs:309]: oh yeah haha, thanks for the snippet, I'll put it tommorrow.

Only whats the reason for`+/- 1` in `to/fromEnum`?

Comment (IAmPara0x) [reply src/BotPlutusInterface/Types.hs:309]: actually in `Data` instances the index of data constructors begins with 1, but in `Enum` it begins with 0. 
Eg: 
```
constrIndex (toConstr $ Error [AnyLog]) == 1
```
and so on... . So, just to be consistent we do (-1) from the index and return the value, and vice-versa.

Comment (zmrocze) [reply src/BotPlutusInterface/Types.hs]: Included this and checked that ordering hasn't changed

**Review by gege251 — APPROVED** [review]
LGTM

**Comment by mikekeke** [comment]
@zmrocze can you transfer this PR from your fork to separate branch of `bpi` repo? So CI checks can run properly.

**Pull request closed without merging**